### ISSUE TITLE
i18n: support variables as translation key

### ISSUE THREAD

**Issue by choeller** [angular/angular#12008]
**I'm submitting a ...**  (check one with "x")

```
[ ] bug report => search github for a similar issue or PR before submitting
[x] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior** 

I currently can't find a way to use variable values as translation keys. I'm basically trying to do something like this:

```
        <select id="state" name="state" [(ngModel)]="task.state" class="form-control">
                <option *ngFor="let state of states" [value]="state" i18n>
                   {{state}}
                </option>
        </select>
```

where states is something like

```
states = ['BACKLOG', 'IN_PROGRESS', 'COMPLETED']
```

**Expected behavior**

It needs to be possible to use calculated values / variables as keys for translation

**Reproduction of the problem**

**What is the motivation / use case for changing the behavior?**

To be able to dynamically translate strings
- **Angular version:** 2.0.1 
  
  
- **Browser:** [all ] 
- **Language:** [all]

**Comment by vicb** [comment]
If you need to translate strings from your ts code, there is already an issue that you can up vote and close this one. Thanks.

**Comment by choeller** [comment]
@vicb I'm not !00% sure, but I think these are two different use-cases. In this case I just want to use the value of a variable as a translation key like demonstrated in the listing. Looking at https://github.com/angular/angular/issues/11405 I would first have to create some additional object mapping my option-value to the translated value like this:
```
this.displayStates = {};
for(let state of states) {
  this.displayStates[state] = translateService.translate(state);
}
```
and then use this map in the template. This is quite wordy and it would be way cooler to be able to just use the state-value as translation-key in the template. 
https://github.com/angular/angular/issues/11405 is also really important for us, but for this use-case it seems like it's a little overkill to do translation in TS.

**Comment by vicb** [comment]
IMO What you want could be easily achieved once we can translate TS.

**Comment by schmuli** [comment]
Would using the [I18nSelectPipe](https://angular.io/docs/ts/latest/api/common/index/I18nSelectPipe-pipe.html) work for this?

**Comment by fredrikredflag** [comment]
@choeller just out of curiosity, is it even possible to see if `states` is a const? If it's dynamic one couldn't get all keys for it, right?

**Comment by ocombe** [comment]
Yes that's no possible with dynamic variables right now.
Once we can use i18n in your code, you could easily fix your use case, the template would be the same (just remove i18n) and the code would be something like:

```
states = [i18n.__('BACKLOG'), i18n.__('IN_PROGRESS'), i18n.__('COMPLETED')]
```

**Comment by skormel** [comment]
> Yes that's no possible with dynamic variables right now.
> Once we can use i18n in TS code, you could easily fix your use case, the template would be the same (just remove i18n) and the code would be something like:
> 
> states = [this.i18n.__('BACKLOG'), this.i18n.__('IN_PROGRESS'), this.i18n.__('COMPLETED')]

Great news, awating this new functionality.

**Comment by matheo** [comment]
@ocombe Still some months to get i18n right? 
I've just read the Design Doc, no approval yet?

**Comment by ocombe** [comment]
It's blocked until some big changes to the compiler are done, so yes still a few months to wait unfortunately

**Comment by Tryptich** [comment]
It's already aproved ?

**Comment by ocombe** [comment]
Approved yes, but not done

**Comment by Tryptich** [comment]
can i join the team who will build it ?

**Comment by ocombe** [comment]
I don't think so, but you can help by commenting the design doc and giving us ideas about how you imagine it to work. And when there will be a PR, any additional review is beneficial

**Comment by Tryptich** [comment]
Ok.

**Comment by xAqweRx** [comment]
Any changes?

**Comment by Im-Harrison** [comment]
@ocombe, have you got any news for this change? Also could you please provide a link to the design doc, i have been unable to find it myself.

**Comment by ocombe** [comment]
The design doc is obsolete for now, we'll rewrite a new one when we start working on code translations.
It is blocked because we are working on runtime translations which is a prerequisite. Once that part is done, we'll start working on it.

**Comment by ocombe** [comment]
Closing this in favor of #11405
